### PR TITLE
Use socketpair to implement Process.Start redirection

### LINKED ISSUE #34158: Consider using socketpair on Unix for Process.Start redirected stdin/out/err
On Linux and macOS today when we fork a process as part of Process.Start, if RedirectStandardInput/Output/Error are set, we create anonymous pipes and use those for the new process' stdin/out/err.  The resulting file descriptors are then wrapped in FileStreams, which are in turn wrapped in StreamReader/Writers.

A downside of this, however, is that async operations performed on those readers/writers end up doing async-over-sync as part of the FileStream implementation, which just queues work items that do Read/Write as part of ReadAsync/WriteAsync, and that can be problematic for systems that spawn lots of processes and monitor their stdout asynchronously.

There are a few ways to address that performance issue.  One is to enable the epoll-based architecture we have underlying System.Net.Sockets to be used by any type with any relevant kind of file descriptor, but that is a significant effort.  In the meantime, we might be able to achieve just as good a result, maybe better, specifically for Process, by using socketpair instead of pipe, and constructing NetworkStreams with Sockets instead of FileStreams.  This would end up implicitly using the right support under sockets, it would end up providing more efficient implementations of APIs like ReadAsync (which has a non-allocating implementation in NetworkStream), etc.

@tmds, @wfurt, any significant roadblocks to this spring to mind?

### PR DESCRIPTION
**Pull request by stephentoub**
Today on Unix, we create an anonymous pipe via pipe/pipe2 to be used for stdin/stdout/stderr on processes created by Process.Start.  We then wrap the resulting file descriptors with FileStreams to hand out via Process.StandardInput/Output/Error.

This has a few issues, however. Any async operations on the resulting stream (or wrapping stream reader) will actually be async-over-sync, and that in turn means that a) any async read will end up blocking a thread pool thread until it's satisified, and b) the operation isn't cancelable.  The implications of (b) are obvious, and the problem with (a) is that code which launches a bunch of processes and uses BeginOutput/ErrorReadLine or the like will end up blocking a bunch of thread pool threads, e.g. https://github.com/dotnet/tye/blob/6a592c01e0bd98a27d0fe2cc3dece9a1a1a02c01/src/tye/Program.RunCommand.cs#L96-L102

This change replaces the pipe/pipe2 calls with socketpair calls, and instead of wrapping the resulting file descriptors with FileStream, wraps them in Sockets and NetworkStreams.  This gives us the full capabilities of the networking stack, which fully supports asynchronous and cancelable reads and writes.

With this PR, synchronous operations performed on the Process.StandardInput/Output/Error have appx the same performance.  Asynchronous operations performed on those have more overhead (as is generally the case, as there's more set-up work to be done), but are more scalable and cancelable.

Fixes https://github.com/dotnet/runtime/issues/34158
cc: @tmds, @wfurt, @janvorli, @adamsitnik, @eiriktsarpalis, @davidfowl

### CODE REVIEW

File: src/libraries/Native/Unix/System.Native/pal_process.c [hunk 1]
```diff
@@ -293,9 +325,9 @@ int32_t SystemNative_ForkAndExecProcess(const char* filename,
...
     // close-on-exec flag to the pipe file descriptors.
-    if ((redirectStdin  && SystemNative_Pipe(stdinFds,  PAL_O_CLOEXEC) != 0) ||
-        (redirectStdout && SystemNative_Pipe(stdoutFds, PAL_O_CLOEXEC) != 0) ||
-        (redirectStderr && SystemNative_Pipe(stderrFds, PAL_O_CLOEXEC) != 0))
+    if ((redirectStdin  && SocketPair(stdinFds) != 0) ||
+        (redirectStdout && SocketPair(stdoutFds) != 0) ||
+        (redirectStderr && SocketPair(stderrFds) != 0))
     {
         success = false;
         goto done;
```
Comment (lpereira) [inline src/libraries/Native/Unix/System.Native/pal_process.c:333]: Unrelated to this patch (and GitHub doesn't let me comment there, grr), but there's no error recovery for the `pipe2()` call below.

File: src/libraries/Native/Unix/System.Native/pal_process.c [hunk 2]
```diff
@@ -203,6 +202,39 @@ static int SetGroups(uint32_t* userGroups, int32_t userGroupsLength, uint32_t* p
...
+        if (result == 0)
+        {
+            while ((result = fcntl(sv[1], F_SETFD, FD_CLOEXEC)) < 0 && errno == EINTR);
+        }
+
+        if (result != 0)
+        {
+            int tmpErrno = errno;
+            close(sv[0]);
+            close(sv[1]);
```
Comment (lpereira) [inline src/libraries/Native/Unix/System.Native/pal_process.c:229]: Instead of closing `sv[0]` and `sv[1]`, set them to `-1` so that `CloseIfOpen()` can actually close them.  Otherwise, this is a race condition and unrelated file descriptors can be closed.

File: src/libraries/Native/Unix/System.Native/pal_process.c [hunk 3]
```diff
@@ -203,6 +202,39 @@ static int SetGroups(uint32_t* userGroups, int32_t userGroupsLength, uint32_t* p
...
+#ifdef SOCK_CLOEXEC
+    type |= SOCK_CLOEXEC;
+#endif
+
+    while ((result = socketpair(AF_UNIX, type, 0, sv)) < 0 && errno == EINTR);
+
+#ifndef SOCK_CLOEXEC
+    if (result == 0)
+    {
+        while ((result = fcntl(sv[0], F_SETFD, FD_CLOEXEC)) < 0 && errno == EINTR);
```
Comment (lpereira) [inline src/libraries/Native/Unix/System.Native/pal_process.c:219]: For consistency sake, please use `READ_END_OF_PIPE` and `WRITE_END_OF_PIPE` instead of `0` and `1`

**Review by lpereira — CHANGES_REQUESTED** [review]

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs [hunk 4]
```diff
@@ -761,16 +762,15 @@ internal static TimeSpan TicksToTimeSpan(double ticks)
...
         /// <param name="access">The access mode.</param>
         /// <returns>The opened stream.</returns>
-        private static FileStream OpenStream(int fd, FileAccess access)
+        private static Stream OpenStream(int fd, FileAccess access)
         {
             Debug.Assert(fd >= 0);
-            return new FileStream(
-                new SafeFileHandle((IntPtr)fd, ownsHandle: true),
-                access, StreamBufferSize, isAsync: false);
+            var socket = new Socket(new SafeSocketHandle((IntPtr)fd, ownsHandle: true));
```
Comment (lpereira) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:772]: On Mac, it's throwing here, because `Socket._isConnected` is apparently never set to `true` when initializing with the file descriptors from `Interop.Sys.ForkAndExecProcess()`

**Comment by adamsitnik** [comment]
@stephentoub nice!

@antonfirsov this is a new scenario that we will have to test when introducing changes to sockets

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:772]: Yeah, we infer _isConnected based on whether we were able to get a remote peer name... apparently that's not working well on macOS with socketpair.
https://github.com/dotnet/runtime/blob/dda0a5701ef6a3a4e66b223695016c2260c617e0/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L233

File: src/libraries/Native/Unix/System.Native/pal_process.c (hunk 2 above)
Comment (stephentoub) [reply src/libraries/Native/Unix/System.Native/pal_process.c:229]: > set them to -1

Why would we want to set them to -1?  That would overwrite the fds, and CloseIfOpen would have nothing to close.  Don't you mean just removing this if block entirely?

Comment (lpereira) [reply src/libraries/Native/Unix/System.Native/pal_process.c:229]: Oh, yes, you're right. Yes, removing this `if` here would fix the race condition. (Maybe add a comment in its place?)

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs (hunk 4 above)
Comment (wfurt) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:772]: I can take a look. I was planning to follow-up on #34727 anyway.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:772]: Thanks!

Comment (lpereira) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:772]: By definition, a `socketpair()` is always connected. Maybe `SafeSocketHandle` could have an internal constructor with a `bool assumeConnected` parameter?

Comment (stephentoub) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:772]: I'd rather just fix the implementation to correctly infer it.  Sounds like @wfurt's already done that. ;)

File: src/libraries/Native/Unix/System.Native/pal_process.c (hunk 1 above)
Comment (stephentoub) [reply src/libraries/Native/Unix/System.Native/pal_process.c:333]: That's called out in the comment:
https://github.com/dotnet/runtime/pull/34861/files/d102eb5d33272cb95e43c16909ddce0903077433#diff-df5b2051e73600a29fbe0b3230daf7d1R340

**Comment by stephentoub** [comment]
@wfurt, FYI, the use of socketpair is still failing here after your changes, with the created Socket coming back as not connected.

**Comment by wfurt** [comment]
> @wfurt, FYI, the use of socketpair is still failing here after your changes, with the created Socket coming back as not connected.

Is this the failing OSX tests or something else? I can take a look. 
```
Darwin
   System.Diagnostics.Process.Tests: [Long Running Test] 'System.Diagnostics.Tests.ProcessTests.Kill_ExitedNonChildProcess_DoesNotThrow', Elapsed: 00:02:07
   System.Diagnostics.Process.Tests: [Long Running Test] 'System.Diagnostics.Tests.ProcessTests.Kill_ExitedNonChildProcess_DoesNotThrow', Elapsed: 00:04:08
   System.Diagnostics.Process.Tests: [Long Running Test] 'System.Diagnostics.Tests.ProcessTests.Kill_ExitedNonChildProcess_DoesNotThrow', Elapsed: 00:06:08
```

**Comment by stephentoub** [comment]
> Is this the failing OSX tests

Yup, e.g.
```
        Child exception:
          System.IO.IOException: The operation is not allowed on non-connected sockets.
        /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs(58,0): at System.Net.Sockets.NetworkStream..ctor(Socket socket, FileAccess access, Boolean ownsSocket)
        /_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs(773,0): at System.Diagnostics.Process.OpenStream(Int32 fd, FileAccess access)
        /_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs(462,0): at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
        /_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs(1221,0): at System.Diagnostics.Process.Start()
        /_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs(1270,0): at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
        /_/src/libraries/System.CodeDom/src/System/CodeDom/Compiler/Executor.cs(68,0): at System.CodeDom.Compiler.Executor.ExecWaitWithCapture(IntPtr userToken, String cmd, String currentDir, TempFileCollection tempFiles, String& outputName, String& errorName)
        /_/src/libraries/System.CodeDom/src/System/CodeDom/Compiler/Executor.cs(32,0): at System.CodeDom.Compiler.Executor.ExecWaitWithCapture(String cmd, TempFileCollection tempFiles, String& outputName, String& errorName)
        /_/src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/ExecutorTests.cs(124,0): at System.CodeDom.Compiler.Tests.ExecutorTests.<ExecWaitWithCapture_OutputIncludesCurrentDirectory>b__7_0()
        /_/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs(359,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

**Comment by wfurt** [comment]
ah, System.CodeDom. I can see it now and I'll take a look. Interestingly, all System.Diagnostics.Process tests are fine.

**Comment by wfurt** [comment]
The getpeername fails with EINVAL. According to man page, 
```
[EINVAL]           socket has been shut down.
```

I verified that if we do not close the other half of the socket pair in SystemNative_ForkAndExecProcess, that the getpeer() succeeds. 
However, that makes the tests hang and I did not dig deeper. (yet?)

We may consider something like
```c#
@@ -190,8 +192,15 @@ private unsafe Socket(SafeSocketHandle handle, bool loadPropertiesFromHandle)
                     {
                         // Local and Remote EP may be different sizes for something like UDS.
                         bufferLength = buffer.Length;
-                        if (SocketPal.GetPeerName(handle, buffer, ref bufferLength) != SocketError.Success)
+                        SocketError result = SocketPal.GetPeerName(handle, buffer, ref bufferLength);
+                        if (result != SocketError.Success )
                         {
+                            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX) && result == SocketError.InvalidArgument)
+                            {
+                                // If socket is not connected, we would get ENOTCONN.
+                                _isConnected = true;
+                            }
+
                             return;
                         }
```

We still won't be able to get RemoteEndPoint if other side disconnected, but all the System.CodeDom tests are passing with that change. Alternatively, we would need to figure out how to avoid or postpone 
https://github.com/dotnet/runtime/blob/feddac7de43999c3f5f9073205d36fd551d746e1/src/libraries/Native/Unix/System.Native/pal_process.c#L440-L442

**Comment by aik-jahoda** [comment]
@wfurt, there is a failing ping test which is caused by this PR. It is related to execute ping cmd:
https://helix.dot.net/api/2019-06-17/jobs/ebaefe2e-2cb3-4aff-86f0-c0570b332e5d/workitems/System.Net.Ping.Functional.Tests

**Comment by wfurt** [comment]
yes, that is because starting process with redirect is busted. I verified that the proposed change in Socket.cs fixes ping tests as well. At this point, that may be best option.

**Comment by stephentoub** [comment]
Thanks for the suggestion, @wfurt.  I'll try that.  Regardless of this change, I think it's important that the Socket(SafeSocketHandle) ctor be usable with socketpair, and in general it's going to be used like this.

**Comment by wfurt** [comment]
> Thanks for the suggestion, @wfurt. I'll try that. Regardless of this change, I think it's important that the Socket(SafeSocketHandle) ctor be usable with socketpair, and in general it's going to be used like this.

There is socketpair test but not with pair to different process. Maybe something to improve. 

I was also thinking about it more we claim socket not to be connected if we fail to get remote address. (with the exception above)
Maybe we can change that logic and assume it is connected unless getpeername() fails with ENOTCONN. That would cover cases when it for example fails with EOPNOTSUPP, ENOTSOCK or any other reason we did not see yet. Lack of RemoteEndPoint may be mildly annoying in that case but the socket would be usable for IO.

**Comment by stephentoub** [comment]
> I was also thinking about it more we claim socket not to be connected if we fail to get remote address. (with the exception above) Maybe we can change that logic and assume it is connected unless getpeername() fails with ENOTCONN. That would cover cases when it for example fails with EOPNOTSUPP, ENOTSOCK or any other reason we did not see yet. Lack of RemoteEndPoint may be mildly annoying in that case but the socket would be usable for IO.

Sounds reasonable.  If we can support a wider array of situations well while not negatively impacting others, I'm all for it.
